Utils/Dataflow/pyDKB/common/custom_readline.py

```diff
@@ -10,12 +10,20 @@
 import fcntl
 
 
-def custom_readline(f, newline):
+def custom_readline(f, newline, markers={}):
```

Comment:
> General comment to the whole set of changes in this function.
> 
> **Expecting markers only at specific place in the flow is not very safe.**
> 
> If the input stream was corrupted and part of it is missed, it may look like this:
> 
> ```
> <data:proper_message><EOM><data:broken_message><EOB>
> ```
> 
> Here's some illustration of how it looks like:
> 
> ```
> ~$ cat test_readline.py
> #!/usr/bin/env python
> 
> import sys
> 
> from pyDKB.common import custom_readline
> 
> EOM='\n'
> EOB='\0'
> 
> for d in custom_readline(sys.stdin, EOM, {'eob': EOB}):
>     print 'd: %r' % d
> ~$ data='{"proper": "message"}\n{"broken_message":\0'
> ~$ echo -ne $data
> {"proper": "message"}
> {"broken_message":~$
> ~$ echo -en $data | hexdump
> 0000000 227b 7270 706f 7265 3a22 2220 656d 7373 
> 0000010 6761 2265 0a7d 227b 7262 6b6f 6e65 6d5f
> 0000020 7365 6173 6567 3a22 0000
> 0000029
> ~$ echo -en $data | ./test_readline.py
> d: '{"proper": "message"}\n'
> d: '{"broken_message":\x00'
> ~$ { echo -en $data; sleep 10; echo -en $data; } | ./test_readline.py
> d: '{"proper": "message"}\n'
> d: '{"broken_message":\x00{"proper": "message"}\n'
> d: '{"broken_message":\x00'
> ```
> 
> What's going on here:
> 
> * if the stream was closed after misplaced `<EOB>`, `custom_readline()` will yield a piece of broken data with `<EOB>`. No one will know it was `<EOB>` -- but it's fine; data are unreadable anyway, stream is closed -- so whatever was read before, it will be processed;
> * if the stream _wasn't_ closed, `custom_readline()` won't return anything until it sees `<EOM>`. But when it does -- the broken message, `<EOB>` and the next proper message will stick together and will be taken as a single message; it won't be decoded (for it is not a properly formatted message) and will be thrown away. So the proper message will be ignored. 

This is true, but the problem here is missed `<EOM>`, not `<EOB>`. Even if we work in `master` that knows nothing about other markers and batch processing, a message being corrupted in a way that erases `<EOM>` will make it "stick" to the next message, and that next message would be ignored, even though it is correct.

> And what's even worse -- until there's`<EOM>` in the input, `custom_readline()` will not return anything at all -- and the whole batch of already read messages will sit in memory and wait, not being processed till the next `<EOB>` or the end of input.

I presume that there is a typo and the second marker is supposed to be `<EOM>` as well, because I don't see any way in which missed `<EOB>` can cause `custom_readline` to hang until next `<EOB>` is encountered. In such case, it's the continuation of the previous text and my answer above still applies. Otherwise - please, clarify this moment.

> What I'd expect in this situation:
> 
> * `<EOB>` is detected;
> * user (or log file) is informed that input stream looks corrupted (we do not expect any data right before `<EOB>`);
> * `<EOB>` is processed in a regular way.

This solves a problem of missing a marker due to it being caught between corrupted messages, but creates another one - what if a sequence of characters inside a (normal or corrupted) message matches some marker's value by accident?

***

All of this is a part of a bigger question "how to deal with errors in communication between worker and supervisor" - or, to be more precise, communication of markers (and losing a marker can be much more dangerous than losing a message). This problem is not unique to batch mode. For example, if supervisor sends a message but `<EOM>` gets lost/corrupted then we have a deadlock: supervisor expects `<EOP>` and worker expects `<EOM>` [1]. Which means that supervisor should have means of breaking such deadlocks, by checking worker's state or something else.

All things considered, I'm not sure that looking for markers inside messages is a solution and that this question fits into this PR. By the way, code in this PR should actually deal with missed `<EOB>` correctly (given that supervisor continues to operate correctly after sending mangled `<EOB>`):

Normal work:
- Supervisor: `message`
- Worker: adds message to batch, batch is too small to process: `<BNC>`
- Supervisor: has no more messages: `<EOB>`
- Worker: processes batch: `messages` + `EOP`
- End

Error:
- Supervisor: `message`
- Worker: adds message to batch, batch is too small to process: `<BNC>`
- Supervisor: has no more messages but error occurs: `something which may or may not include <EOB> or its pieces somewhere`
- Worker: I got something. It's not a known marker, so it's a message. It's a faulty message, so I discard it. My batch is still too small to process: `<BNC>`
- Supervisor: has no more messages: `<EOB>`
- Worker: processes batch: `messages` + `EOP`
- End

[1] This example can be simulated by starting a stage in terminal mode with standard `<EOM>`=`\n` and forgetting to press ENTER after a message. One can look at the screen and realize their mistake, but hey, that's some advanced supervisor technique...
 
> **Returning "name of marker" instead of the marker itself may be ambiguous.**
> 
> Whenever the function is called from, the context is aware of marker values and can compare returned value to them. On the other hand, sequence of `eob<EOF>` (with proper `EOF` == "STDIN is closed"), that originally would be processed as "unexpected end of stream" by the `InputStream`, now will be taken as `<EOB>` at the end of stream:
> 
> ```
> ~$ data1='eob'
> ~$ data2='\0'
> ~$ echo -ne $data1 | ./test_readline.py
> d: 'eob'
> ~$ echo -ne $data2 | ./test_readline.py
> d: 'eob'
> d: ''
> ```
> 
> _(BTW, the extra empty line is an unexpected result)_
> 
> Maybe it's not a big problem -- `<EOB>` at the end of stream won't make worker to do something unexpected -- but it is just an example of possibility of incorrect interpretation of a plain string `'eob'` as a control marker.

Fair point, will do.

Reply:
@Evildoor
> > **Expecting markers only at specific place in the flow is not very safe.**
> > If the input stream was corrupted and part of it is missed, it may look like this:
> > ```
> > <data:proper_message><EOM><data:broken_message><EOB>
> > ```
> > 
> > * if the stream _wasn't_ closed, `custom_readline()` won't return anything until it sees `<EOM>`. But when it does -- the broken message, `<EOB>` and the next proper message will stick together and will be taken as a single message; it won't be decoded (for it is not a properly formatted message) and will be thrown away. So the proper message will be ignored.
> 
> This is true, but the problem here is missed `<EOM>`, not `<EOB>`.

The problem is not in "missed `<EOM>`" -- some issues in the data transfer *may* happen, and it is not what we can control with the protocol specification and encoder/decoder implementation. My point is that in the current implementation of parser (how the presence of `<EOB>` is checked) makes this kind of issue more harmful than it could be.

> Even if we work in `master` that knows nothing about other markers and batch processing, a message being corrupted in a way that erases `<EOM>` will make it "stick" to the next message, and that next message would be ignored, even though it is correct.

Right, the correct message following the broken one will be thrown away. But in this case:
* the flow looks this way: `<data_broken+data_correct><EOM>`: there's no way to separate broken data from correct (while with `<EOB>` between them it is possible);
* the stage *does* process all the messages prior to the broken one and stuck waiting for the next `<EOM>` with only the broken one in memory, not the whole batch; in other words -- everything that can be processed, is processed without delay.

> > And what's even worse -- until there's`<EOM>` in the input, `custom_readline()` will not return anything at all -- and the whole batch of already read messages will sit in memory and wait, not being processed till the next `<EOB>` or the end of input.
> 
> I presume that there is a typo and the second marker is supposed to be `<EOM>` as well,

No, it's `<EOB>`.

> because I don't see any way in which missed `<EOB>` can cause `custom_readline` to hang until next `<EOB>` is encountered.

Not the `custom_readline()` in particular, but the stage itself; the buffer of messages, for now, is a stage attribute, not the `custom_readline()`'s.

> In such case, it's the continuation of the previous text and my answer above still applies. Otherwise - please, clarify this moment.

Here you are: the _stage_ operating in the batch mode (even in worker-driven scenario) will not process anything until it:
* gets enough messages in buffer;
* gets `<EOB>`;
* gets input stream closed.

If supervisor sent `<EOB>`, keeping the stream open but having no more messages for processing for the next hour, the worker will sit and wait for the whole hour instead of processing the messages it already has.
*Just in case, reminder: we're talking about situation with the broken message (without `<EOM>`) prior to the `<EOB>`, not the general situation.*

> > What I'd expect in this situation:
> > 
> > * `<EOB>` is detected;
> > * user (or log file) is informed that input stream looks corrupted (we do not expect any data right before `<EOB>`);
> > * `<EOB>` is processed in a regular way.
> 
> This solves a problem of missing a marker due to it being caught between corrupted messages, but creates another one - what if a sequence of characters inside a (normal or corrupted) message matches some marker's value by accident?

Ahha, why do you think `<EOM>` is a NULL ASCII character, not plain "eom"? ;)
It's a well known issue: everything that has special meaning must be properly treated in the data (just as registered keyword "table" in SQL language, for example).
The simpliest way to get rid of this is to encode data with base64 and make sure none of these 64 ASCII symbols are used as special ones.

> All of this is a part of a bigger question "how to deal with errors in communication between worker and supervisor" - or, to be more precise, communication of markers (and losing a marker can be much more dangerous than losing a message). This problem is not unique to batch mode. For example, if supervisor sends a message but `<EOM>` gets lost/corrupted then we have a deadlock: supervisor expects `<EOP>` and worker expects `<EOM>` [1]. Which means that supervisor should have means of breaking such deadlocks, by checking worker's state or something else.

Right, this should be improved. Feel free to write it down somewhere (Google docs with the protocol description, Trello cards, your own notepad), add your ideas on how it can be solved to make sure you won't forget when decide to take care of it.

But here and now your main objective is *not* to detect and solve *existing* issues; it is to introduce new functionality without inviting new ones where it can be helped. Existing issue is not an excuse to add new ones saying "why bother now, if it's not working anyway; one day someone will take care of this, so let him/her take care of all at once".

> All things considered, I'm not sure that looking for markers inside messages is a solution and that this question fits into this PR.

I do insist.

> By the way, code in this PR should actually deal with missed `<EOB>` correctly (given that supervisor continues to operate correctly after sending mangled `<EOB>`):
> 
> Normal work:
> 
> * Supervisor: `message`
> * Worker: adds message to batch, batch is too small to process: `<BNC>`
> * Supervisor: has no more messages: `<EOB>`
> * Worker: processes batch: `messages` + `EOP`
> * End
> 
> Error:
> 
> * Supervisor: `message`
> * Worker: adds message to batch, batch is too small to process: `<BNC>`
> * Supervisor: has no more messages but error occurs: `something which may or may not include <EOB> or its pieces somewhere`
> * Worker: I got something. It's not a known marker, so it's a message.

Wrong: until there's `<EOM>`, it's not a message either. So I will wait for `<EOM>` (for how long? a minute? an hour?) until decide that...

> It's a faulty message, so I discard it.
> My batch is still too small to process: `<BNC>`
> * Supervisor: has no more messages: `<EOB>`

Wrong: if worker saw `<EOM>` finally came, it's very likely that supervisor has a new pack of messages, in which case worker will have some messages and maybe even fill the batch, or -- as you suggested -- will see another `<EOB>`.

> * Worker: processes batch: `messages` + `EOP`
> * End
> 
> [1] This example can be simulated by starting a stage in terminal mode with standard `<EOM>`=`\n` and forgetting to press ENTER after a message. One can look at the screen and realize their mistake, but hey, that's some advanced supervisor technique...

I don't think we should continue the discussion for it already took too much words, so here's my ultimate resolution on this question: I won't accept the PR until the code can recognize `<EOB>` in data stream independently of the marker's position.

---

> > **Returning "name of marker" instead of the marker itself may be ambiguous.**
> 
> Fair point, will do.

Thank you.

Reply:
Understood, will do.